eskip/lexer.go

```diff
@@ -17,218 +17,306 @@ package eskip
+func isSymbolChar(c byte) bool  { return isUnderscore(c) || isAlpha(c) || isDigit(c) }
+func isDecimalChar(c byte) bool { return c == decimalChar }
+func isNumberChar(c byte) bool  { return isDecimalChar(c) || isDigit(c) }
+
+func scanWhile(code string, p charPredicate) ([]byte, string) {
+	var b []byte
+	for len(code) > 0 && p(code[0]) {
+		b = append(b, code[0])
+		code = code[1:]
+	}
+
+	return b, code
+}
+
+func scanVoid(code string, p charPredicate) string {
+	_, rest := scanWhile(code, p)
+	return rest
+}
+
+func scanEscaped(delimiter byte, code string) ([]byte, string) {
+	var b []byte
 	escaped := false
-	for i := 0; i < len(s); i++ {
-		c := s[i : i+1]
-		switch {
-		case escaped && strings.Index(chars, c) >= 0:
-			r = append(r, c)
-			escaped = false
-		case escaped:
-			r = append(r, "\\", c)
+	for len(code) > 0 {
+		c := code[0]
+		isDelimiter := c == delimiter
+		isEscapeChar := c == escapeChar
+
+		if escaped {
+			if !isDelimiter && !isEscapeChar {
+				b = append(b, escapeChar)
+			}
+
+			b = append(b, c)
 			escaped = false
-		case c == "\\":
-			escaped = true
-		default:
-			r = append(r, c)
+		} else {
+			if isDelimiter {
+				return b, code
+			}
+
+			if isEscapeChar {
+				escaped = true
+			} else {
+				b = append(b, c)
+			}
 		}
+
+		code = code[1:]
+	}
+
+	return b, code
+}
+
+func scanRegexpLiteral(code string) (t token, rest string, err error) {
+	b, rest := scanEscaped('/', code[1:])
+	if len(rest) == 0 {
+		err = incompleteToken
+		return
 	}
 
-	return strings.Join(r, "")
+	rest = rest[1:]
+	t.id = regexpliteral
+	t.val = string(b)
+	return
 }
 
-// conversion error ignored, tokenizer expression already checked format
-func convertNumber(s string) float64 {
-	n, _ := strconv.ParseFloat(s, 64)
-	return n
+func scanRegexpOrComment(code string) (t token, rest string, err error) {
+	if len(code) < 2 {
+		rest = code
+		err = invalidCharacter
+		return
+	}
+
+	if code[1] == '/' {
+		rest = scanComment(code)
+		err = void
+		return
+	}
+
+	t, rest, err = scanRegexpLiteral(code)
+	return
 }
 
-// unescaping only '"' and '\'
-func convertString(s string) string {
-	if s[0:1] == "`" {
-		return s[1 : len(s)-1]
+func scanStringLiteral(delimiter byte, code string) (t token, rest string, err error) {
+	b, rest := scanEscaped(delimiter, code[1:])
+	if len(rest) == 0 {
+		err = incompleteToken
+		return
 	}
 
-	return unescape(s[1:len(s)-1], "\\\"")
+	rest = rest[1:]
+	t.id = stringliteral
+	t.val = string(b)
+	return
 }
 
-// unescaping only '/'
-func convertRegexp(s string) string {
-	return unescape(s[1:len(s)-1], "/")
+func scanWhitespace(code string) string { return scanVoid(code, isWhitespace) }
+func scanComment(code string) string {
+	return scanVoid(code, func(c byte) bool { return !isNewline(c) })
 }
+func scanStringLiteral1(code string) (token, string, error) { return scanStringLiteral('"', code) }
+func scanStringLiteral2(code string) (token, string, error) { return scanStringLiteral('`', code) }
```

Comment:
the names scanStringLiteralDoubleQuote and scanStringLiteralBacktick felt just so.. something. Do you have a good late-wednesday suggestion?